### PR TITLE
Add RustAdapter (#2892)

### PR DESCRIPTION
**Pull request by ksanth850**
(cherry picked from commit b3018d54feeec44daf2c615ce22cd2cee2052cc8)

Add rust adapter process for outofprocess mode

(cherry picked from commit 976bad478ef78e789c734a50fd61d1b9c8805cc3)

improve rust remote startup error handling

(cherry picked from commit c96cd6df52dfc3f0f503c9b6b73430c412413610)

changes to support remote debug

(cherry picked from commit 0b98bb734c3645f0b24051a0a61fe732d3b526fd)

remote send_to mut hack

(cherry picked from commit cd35c300a1a3146500f5b445990e767b2dd2c450)

rename rust_adapter_process to WPEHost

(cherry picked from commit 81ce0f9f06cd981366450bf5634b655ef4691f49)

move rust WPEHost from rdkservices to thunder_rs

(cherry picked from commit 0044095c19494b7bbb69044f96887092f0d334bf)

Changed C++/Rust interface to exclude the use of void pointers.

(cherry picked from commit ab7888fa66dac52e2756ff31c7a466f501e62f72)

allows RustAdapter to build with older Thunder version

(cherry picked from commit fdec24139023e980e4fe08ad18f2b887245f9d02)

Added capability to override the rust shared library name. Currently code uses the callsign
name.

(cherry picked from commit f2bc1e7cdc883f5e96eb2afa4305f1ec7c4ef7b9)

Added Athorization token to plugin's configuration.

(cherry picked from commit 0824cd19e7f0dfc5fda966570cddf55ef5e70bd3)

Co-authored-by: mrollins <mark_rollins@cable.comcast.com>
Co-authored-by: Karunakaran A <48997923+karuna2git@users.noreply.github.com>
(cherry picked from commit 06a74bf0ae0b6954807d84354094ba9288ad43b6)

### CODE REVIEW

File: RustAdapter/RustAdapter.cpp [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+{
+  using namespace WPEFramework;
+
+  auto nodeId = Core::NodeId{ "/tmp/SecurityAgent/token" };
+  auto engine = Core::ProxyType<RPC::InvokeServerType<1, 0, 4>>::Create();
+  auto client = Core::ProxyType<RPC::CommunicatorClient>::Create(nodeId, Core::ProxyType<Core::IIPCServer>(engine));
+
+  int result = -1;
+  if ((client.IsValid() == true) && (client->IsOpen() == false)) {
+    PluginHost::IAuthenticate* securityAgentInterface = client->Open<PluginHost::IAuthenticate>("SecurityAgent");
```
Comment (npoltorapavlo) [inline RustAdapter/RustAdapter.cpp:171]: this violates `7. Inter-plugin communication` of [readme](https://github.com/rdkcentral/rdkservices#readme)
please use QueryInterfaceByCallsign

**Review by npoltorapavlo — CHANGES_REQUESTED** [review]

**Review by karuna2git — APPROVED** [review]

File: RustAdapter/RustAdapter.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  int n = snprintf(buff, sizeof(buff), "plugin://%s", callsign.c_str());
+  buff[255] = '\0';
+
+  auto auth = service->QueryInterfaceByCallsign<WPEFramework::PluginHost::IAuthenticate>("SecurityAgent");
+  if (auth != nullptr) {
+    string encoded;
+    if (auth->CreateToken(n, (const uint8_t*) buff, encoded) == WPEFramework::Core::ERROR_NONE) {
+      return encoded;
+    }
+  }
```
Comment (npoltorapavlo) [inline RustAdapter/RustAdapter.cpp:174]: should also release the interface `auth->Release();`

**Review by npoltorapavlo — CHANGES_REQUESTED** [review]

File: RustAdapter/RustAdapter.cpp (hunk 2 above)
Comment (karuna2git) [reply RustAdapter/RustAdapter.cpp]: added `auth->Release();`

File: RustAdapter/RustAdapter.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  int n = snprintf(buff, sizeof(buff), "plugin://%s", callsign.c_str());
+  buff[255] = '\0';
+
+  auto auth = service->QueryInterfaceByCallsign<WPEFramework::PluginHost::IAuthenticate>("SecurityAgent");
+  if (auth != nullptr) {
+    string encoded;
+    if (auth->CreateToken(n, (const uint8_t*) buff, encoded) == WPEFramework::Core::ERROR_NONE) {
+      return encoded;
+    }
+    auth->Release();
```
Comment (npoltorapavlo) [inline RustAdapter/RustAdapter.cpp:174]: please replace `return encoded;` with`auth_token = encoded;`

**Review by npoltorapavlo — CHANGES_REQUESTED** [review]

File: RustAdapter/RustAdapter.cpp (hunk 3 above)
Comment (karuna2git) [reply RustAdapter/RustAdapter.cpp]: I overlooked. Thanks for pointing.. its done

**Comment by npoltorapavlo** [comment]
looks good

**Review by karuna2git — APPROVED** [review]

**Pull request merged**